src/ghost_cell.rs

```diff
@@ -311,23 +311,78 @@ impl<'brand, T: ?Sized> GhostCell<'brand, T> {
         _: &'a mut GhostToken<'brand>
     ) -> Option<(&'a mut T, &'a mut Q, &'a mut R)> where T: Sized {
         // we require that `T`, `Q`, `R` are `Sized`, so no fat pointer problems.
-        if     core::ptr::eq(cell_a, cell_b as *const _ as *const _) 
-            || core::ptr::eq(cell_b, cell_c as *const _ as *const _) 
-            || core::ptr::eq(cell_c, cell_a as *const _ as *const _) 
-        {
-            None
-        } else {
+        check_distinct([cell_a as *const _ as *const (), cell_b as *const _ as *const (), cell_c as *const _ as *const ()])?;
+        unsafe {
+            Some((
+                &mut *cell_a.value.get(),
+                &mut *cell_b.value.get(),
+                &mut *cell_c.value.get()
+            ))
+        }
+    }
+}
+
+#[cfg(feature = "experimental-multiple-mutable-borrows")]
+/// Returns `Some(())` if the inputs are distinct, and `None` otherwise.
+fn check_distinct<const N: usize>(arr: [*const (); N]) -> Option<()> {
+    for i in 0..N {
+        for j in i+1..N {
+            if core::ptr::eq(arr[i], arr[j]) {
+                return None;
+            }
+        }
+    }
+    Some(())
+    // TODO: if the array is large enough, sort the values instead.
```

Comment:
That's a good point, sorting is probably less efficient on small arrays.

Do you have a rough idea of where the tipping point would be?